azure/defaults_test.go

```diff
@@ -24,11 +24,109 @@ import (
 	"sync"
 	"testing"
 
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/cloud"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime"
 	"github.com/Azure/go-autorest/autorest"
 	. "github.com/onsi/gomega"
 	"sigs.k8s.io/cluster-api-provider-azure/util/tele"
 )
 
+// TestARMClientOptions tests the `ARMClientOptions()` factory function.
+func TestARMClientOptions(t *testing.T) {
+	g := NewWithT(t)
+
+	tests := []struct {
+		name          string
+		cloudName     string
+		expectedCloud cloud.Configuration
+		expectError   bool
+	}{
+		{
+			name:          "should return default client options if cloudName is empty",
+			cloudName:     "",
+			expectedCloud: cloud.Configuration{},
+		},
+		{
+			name:          "should return Azure public cloud client options",
+			cloudName:     PublicCloudName,
+			expectedCloud: cloud.AzurePublic,
+		},
+		{
+			name:          "should return Azure China cloud client options",
+			cloudName:     ChinaCloudName,
+			expectedCloud: cloud.AzureChina,
+		},
+		{
+			name:          "should return Azure government cloud client options",
+			cloudName:     USGovernmentCloudName,
+			expectedCloud: cloud.AzureGovernment,
+		},
+		{
+			name:        "should return error if cloudName is unrecognized",
+			cloudName:   "AzureUnrecognizedCloud",
+			expectError: true,
+		},
+	}
+	for _, tc := range tests {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			opts, err := ARMClientOptions(tc.cloudName)
+			if tc.expectError {
+				g.Expect(err).To(HaveOccurred())
+				return
+			}
+			g.Expect(err).NotTo(HaveOccurred())
+			g.Expect(opts.Cloud).To(Equal(tc.expectedCloud))
+			g.Expect(opts.Retry.MaxRetries).To(BeNumerically("==", -1))
+			g.Expect(opts.PerCallPolicies).To(HaveLen(2))
+		})
+	}
+}
+
+// TestPerCallPolicies tests the per-call policies returned by `ARMClientOptions()`.
+func TestPerCallPolicies(t *testing.T) {
+	g := NewWithT(t)
+
+	corrID := "test-1234abcd-5678efgh"
+	// This server will check that the correlation ID and user-agent are set correctly.
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		g.Expect(r.Header.Get("User-Agent")).To(ContainSubstring("cluster-api-provider-azure/"))
+		g.Expect(r.Header.Get(string(tele.CorrIDKeyVal))).To(Equal(corrID))
+		fmt.Fprintf(w, "Hello, %s", r.Proto)
+	}))
+	defer server.Close()
+
+	// Call the factory function and ensure it has both PerCallPolicies.
+	opts, err := ARMClientOptions("")
+	g.Expect(err).NotTo(HaveOccurred())
+	g.Expect(opts.PerCallPolicies).To(HaveLen(2))
+	g.Expect(opts.PerCallPolicies).To(ContainElement(BeAssignableToTypeOf(correlationIDPolicy{})))
```

Comment:
TIL you can put other matchers inside `ContainElement`. It's not as readable as most Ginkgo code, but it works as desired: does it contain an element of this type?